File: backend/src/application/use-cases/CreateHelpRequest.ts
import { HelpRequest, HelpRequestType, HelpRequestPriority } from '../../domain/entities/HelpRequest';
import { IHelpRequestRepository } from '../../domain/repositories/IHelpRequestRepository';
import { IEmergencyRepository } from '../../domain/repositories/IEmergencyRepository';
import { IHelpOfferRepository } from '../../domain/repositories/IHelpOfferRepository';
import { MatchingService } from '../../domain/services/MatchingService';

export interface CreateHelpRequestDTO {
  emergencyId: string;
  requesterId: string;
  type: HelpRequestType;
  priority: HelpRequestPriority;
  title: string;
  description: string;
  latitude: number;
  longitude: number;
  numberOfPeople: number;
}

export class CreateHelpRequest {
  private matchingService: MatchingService;

  constructor(
    private readonly helpRequestRepository: IHelpRequestRepository,
    private readonly emergencyRepository: IEmergencyRepository,
    private readonly helpOfferRepository: IHelpOfferRepository
  ) {
    this.matchingService = new MatchingService();
  }

  async execute(dto: CreateHelpRequestDTO): Promise<{
    request: HelpRequest;
    suggestedMatches: Array<{ offerId: string; distanceKm: number; score: number }>;
  }> {
    // Verify emergency exists and is active
    const emergency = await this.emergencyRepository.findById(dto.emergencyId);
    if (!emergency) {
      throw new Error('Emergency not found');
    }
    if (!emergency.isActive()) {
      throw new Error('Emergency is no longer active');
    }

    // Create the request
    const request = new HelpRequest({
      emergencyId: dto.emergencyId,
      requesterId: dto.requesterId,
      type: dto.type,
      priority: dto.priority,
      title: dto.title,
      description: dto.description,
      latitude: dto.latitude,
      longitude: dto.longitude,
      numberOfPeople: dto.numberOfPeople,
    });

    const saved = await this.helpRequestRepository.save(request);

    // Find potential matches
    const availableOffers = await this.helpOfferRepository.findAvailableByType(
      dto.emergencyId,
      dto.type
    );

    const matches = this.matchingService.findBestMatches(saved, availableOffers, 5);

    return {
      request: saved,
      suggestedMatches: matches.map((m) => ({
        offerId: m.offer.id,
        distanceKm: Math.round(m.distanceKm * 10) / 10,
        score: Math.round(m.score),
      })),
    };
  }
}

File: backend-dotnet/src/Hestia.Application/UseCases/CreateHelpRequest.cs
using Hestia.Application.DTOs;
using Hestia.Domain.Entities;
using Hestia.Domain.Repositories;
using Hestia.Domain.Services;

namespace Hestia.Application.UseCases;

public class CreateHelpRequest
{
    private readonly IHelpRequestRepository _helpRequestRepository;
    private readonly IEmergencyRepository _emergencyRepository;
    private readonly IHelpOfferRepository _helpOfferRepository;
    private readonly MatchingService _matchingService;

    public CreateHelpRequest(
        IHelpRequestRepository helpRequestRepository,
        IEmergencyRepository emergencyRepository,
        IHelpOfferRepository helpOfferRepository,
        MatchingService matchingService)
    {
        _helpRequestRepository = helpRequestRepository;
        _emergencyRepository = emergencyRepository;
        _helpOfferRepository = helpOfferRepository;
        _matchingService = matchingService;
    }

    public async Task<CreateHelpRequestResponse> ExecuteAsync(
        CreateHelpRequestRequest dto,
        Guid requesterId,
        CancellationToken ct = default)
    {
        var emergency = await _emergencyRepository.FindByIdAsync(dto.EmergencyId, ct)
            ?? throw new ApplicationException("Emergency not found");

        if (!emergency.IsActive())
            throw new ApplicationException("Emergency is no longer active");

        var request = new HelpRequest(
            dto.EmergencyId, requesterId, dto.Type, dto.Priority,
            dto.Title, dto.Description,
            dto.Latitude, dto.Longitude, dto.NumberOfPeople);

        var saved = await _helpRequestRepository.SaveAsync(request, ct);

        var availableOffers = await _helpOfferRepository.FindAvailableByTypeAsync(dto.EmergencyId, dto.Type, ct);
        var matches = _matchingService.FindBestMatches(saved, availableOffers, 5);

        return new CreateHelpRequestResponse
        {
            Request = saved,
            SuggestedMatches = matches.Select(m => new SuggestedMatch
            {
                OfferId = m.Offer.Id,
                DistanceKm = Math.Round(m.DistanceKm, 1),
                Score = Math.Round(m.Score)
            }).ToList()
        };
    }
}
